File: Sources/SetlistCore/LibraryStore.swift
import Foundation

/// Where the library lives between launches. Phase A keeps everything on the
/// iPad, so the only implementation writes a file inside the app's own
/// container — a location the user can delete by deleting the app.
public protocol LibraryStore {
    func load() throws -> Library
    func save(_ library: Library) throws
}

public struct FileLibraryStore: LibraryStore {
    public let url: URL

    public init(url: URL) {
        self.url = url
    }

    /// The store's home inside a container directory. Named, not defaulted, so
    /// a caller cannot accidentally write outside the container.
    public static func inContainer(_ container: URL) -> FileLibraryStore {
        FileLibraryStore(url: container.appendingPathComponent("library.json"))
    }

    public func load() throws -> Library {
        guard FileManager.default.fileExists(atPath: url.path) else { return Library() }
        return try JSONDecoder().decode(Library.self, from: Data(contentsOf: url))
    }

    public func save(_ library: Library) throws {
        try FileManager.default.createDirectory(
            at: url.deletingLastPathComponent(),
            withIntermediateDirectories: true
        )
        let data = try JSONEncoder().encode(library)
        // Atomic: a set half-written while the app is killed on stage would
        // lose the material, and material loss is the one unforgivable bug.
        try data.write(to: url, options: .atomic)
    }
}

File: packages/setlist_core/lib/src/library_store.dart
import 'dart:convert';
import 'dart:io';

import 'material.dart';

/// Where the library lives between launches.
///
/// Phase A keeps everything on the device, so the only implementation writes a
/// file inside the app's own container — a location the user can delete by
/// deleting the app.
abstract interface class LibraryStore {
  Future<Library> load();
  Future<void> save(Library library);
}

class FileLibraryStore implements LibraryStore {
  const FileLibraryStore(this.file);

  /// The store's home inside a container directory. Named, not defaulted, so a
  /// caller cannot accidentally write outside the container.
  factory FileLibraryStore.inContainer(Directory container) =>
      FileLibraryStore(File('${container.path}/library.json'));

  final File file;

  @override
  Future<Library> load() async {
    if (!await file.exists()) return Library();
    return Library.fromJson(json.decode(await file.readAsString()) as Map<String, dynamic>);
  }

  @override
  Future<void> save(Library library) async {
    await file.parent.create(recursive: true);
    // Written aside and moved into place: a set half-written while the app is
    // killed on stage would lose the material, and losing material is the one
    // unforgivable bug. A rename within a directory is atomic.
    final pending = File('${file.path}.writing');
    await pending.writeAsString(json.encode(library.toJson()), flush: true);
    await pending.rename(file.path);
  }
}
